### ISSUE TITLE
TypeError: $tw.utils.replaceString is not a function in tiddlywebadaptor.js

### ISSUE THREAD

**Issue by evadeflow** [Jermolene/TiddlyWiki5#2695]
I have a 'Node-ified' folder of tiddlers laid out like this:

```
$ tree -L 1
.
├── bin
├── tiddlers
├── TiddlyWiki5
└── tiddlywiki.info
```

The `TiddlyWiki5` folder is a git submodule that injects the TW sources into my tree:

```
$ cat .gitmodules 
[submodule "TiddlyWiki5"]
        path = TiddlyWiki5
        url = https://github.com/Jermolene/TiddlyWiki5.git
```

(The reason I'm experimenting with this setup is that I'd like to be able to install everything needed for working with TW by simply cloning my tiddler repo.)

When I run the following command:

```
$ ./bin/node-linux64 TiddlyWiki5/tiddlywiki.js --server
Serving on 127.0.0.1:8080
(press ctrl-C to exit)
syncer-server: Dispatching 'save' task: $:/StoryList
FileSystem: Saved file /home/dwolfe/projects/reverie/tiddlers/$__StoryList.tid
```

and then visit `localhost:8080` in my browser, I get an error page saying:

```
                                     Internal JavaScript Error

Well, this is embarrassing. It is recommended that you restart TiddlyWiki by refreshing your browser

                        TypeError: $tw.utils.replaceString is not a function
```

The file `bin/node-linux64` is a copy of the Node binary from [here](https://nodejs.org/download/release/latest-v7.x/node-v7.3.0-linux-x64.tar.gz) that I've included in my tiddler repo as part of my "just clone & go" experiment. But I get the same error if I use a system-installed `node` binary instead, so I don't think it's related to this problem.

One final data point (which I hope will be useful to TW devs). If I remove the `tiddlyweb` plugin like this:

```diff
diff --git a/tiddlywiki.info b/tiddlywiki.info
index 8e326b8..9a8c71e 100644
--- a/tiddlywiki.info
+++ b/tiddlywiki.info
@@ -1,7 +1,6 @@
 {
     "description": "Basic client-server edition",
     "plugins": [
-        "tiddlywiki/tiddlyweb",
         "tiddlywiki/filesystem",
         "tiddlywiki/highlight"
     ],
@@ -55,4 +54,4 @@
             "text/plain"
         ]
     }
-}
\ No newline at end of file
+}
```

I get a warning when I start `node`:

```
$ ./bin/node-linux64 TiddlyWiki5/tiddlywiki.js --server
Serving on 127.0.0.1:8080
(press ctrl-C to exit)
Warning: Plugins required for client-server operation ("tiddlywiki/filesystem" and "tiddlywiki/tiddlyweb") are missing from tiddlywiki.info file
syncer-server: Dispatching 'save' task: $:/StoryList
FileSystem: Saved file /home/dwolfe/projects/reverie/tiddlers/$__StoryList.tid
```

but... the problem goes away, ~~and everything seems to work, more-or-less~~. Could this be related to [08cfa88](https://github.com/Jermolene/TiddlyWiki5/commit/08cfa88249f12262d0f5d4d371e3d6a5a2cecf1b#diff-a44355b62ef95b5b6faeb20a6050e800)?

I realize my setup may be slightly unusual, and perhaps this isn't really 'supposed' to work, but I thought I'd submit an issue for it, just in case I've uncovered an *actual* bug...

**Comment by evadeflow** [comment]
Erm... this is probably obvious to many, but I was incorrect when I stated that everything works with the tiddlyweb plugin disabled. I can create new tiddlers, but they (silently) fail to be saved, so... definitely not 'working' without tiddlyweb. (Which isn't surprising after having a look at  [tiddlyweb/readme.tid](https://github.com/Jermolene/TiddlyWiki5/blob/master/plugins/tiddlywiki/tiddlyweb/readme.tid)).

**Comment by Jermolene** [comment]
Hi @evadeflow what happens when you run this:

```
cd /home/dwolfe/projects/reverie/TiddlyWiki5
./bin/node-linux64 ./tiddlywiki.js editions/tw5.com --build index
```

If the command runs successfully, could you post `/home/dwolfe/projects/reverie/TiddlyWiki5/editions/tw5.com/output/index.html` to a [gist](https://gist.github.com/)?

**Comment by evadeflow** [comment]
Sure, here you go:

* https://gist.github.com/evadeflow/ec600979cd96697a1044a6744c764a6a

FWIW—after posting that gist—I opened the `index.html` file in a browser. It did *not* exhibit the *Internal Javascript Error* issue. (Oh, and just so you know: I executed exactly the command you suggested, except that I had to change the `node` path to `../bin/node-linux64` .)

**Comment by Arlen22** [comment]
@evadeflow would you mind posting an `ls` of the tiddlers folder? You can remove anything that doesn't start with `$`. Just a shot in the dark.

**Comment by evadeflow** [comment]
Sure thing:

```
$ ls -1 tiddlers/\$*
tiddlers/$__config_DefaultSidebarTab.tid
tiddlers/$__config_PageControlButtons_Visibility_$__core_ui_Buttons_fold-all.tid
tiddlers/$__config_PageControlButtons_Visibility_$__core_ui_Buttons_save-wiki.tid
tiddlers/$__config_PageControlButtons_Visibility_$__core_ui_Buttons_unfold-all.tid
tiddlers/$__config_WikiParserRules_Inline_wikilink.tid
tiddlers/$__core.json.tid
tiddlers/$__DefaultTiddlers.tid
tiddlers/$__pluginlibrary_tobibeer_official.tid
tiddlers/$__plugins_tiddlywiki_highlight.json.tid
tiddlers/$__plugins_tiddlywiki_markdown.json.tid
tiddlers/$__plugins_tobibeer_xlist.json.tid
tiddlers/$__SiteSubtitle.tid
tiddlers/$__SiteTitle.tid
tiddlers/$__StoryList.tid
tiddlers/$__themes_tiddlywiki_snowwhite.json.tid
tiddlers/$__themes_tiddlywiki_vanilla.json.tid
```

**Comment by evadeflow** [comment]
I thought I'd add a few more pieces of information that may help in diagnosing this issue. First is that I *am* able to run my wiki normally if I use the `tiddlywiki` wrapper script that's installed when I type `sudo npm install -g tiddlywiki`. In other words, this works fine when run from the root of my working copy:

```
$ tiddlywiki --server
```

The second data point is: the [`index.html`](https://gist.github.com/evadeflow/ec600979cd96697a1044a6744c764a6a) file Jeremy had me create yesterday works fine when I attempt to run it in the same way that's causing *my* wiki to generate an error. I started by creating an empty wiki:

```
$ tiddlywiki tw5.com --init
Copied edition 'empty' to tw5.com
```

Then, I added the filesystem and tiddlyweb plugins to the `tw5.com/tiddlywiki.info` created by `--init`. Conceptually, this amounted to applying this patch:

```diff
--- tw5.com/tiddlywiki.info.orig        2017-01-05 13:42:41.542325772 -0500
+++ tw5.com/tiddlywiki.info     2017-01-05 13:43:10.342325772 -0500
@@ -1,6 +1,9 @@
 {
     "description": "Empty edition",
-    "plugins": [],
+    "plugins": [
+        "tiddlywiki/filesystem",
+        "tiddlywiki/tiddlyweb"
+    ],
     "themes": [
         "tiddlywiki/vanilla",
         "tiddlywiki/snowwhite"
@@ -51,4 +54,4 @@
             "text/plain"
         ]
     }
-}
\ No newline at end of file
+}
```

Then I ran:

```
$ tiddlywiki tw5.com --load index.html --server
```

and hit `localhost:8080` in a browser. No problems at all, everything worked great. After that, I killed the server and ran:

```
$ ../reverie/bin/node-linux64 ../reverie/TiddlyWiki5/tiddlywiki.js --server
Serving on 127.0.0.1:8080
(press ctrl-C to exit)
syncer-server: Dispatching 'save' task: $:/StoryList
FileSystem: Saved file /home/dwolfe/projects/tw5.com/tiddlers/$__StoryList.tid
```

Again, everything worked just fine. So... it seems that my wiki is somehow b0rk3d. I would *really* like to know:

1. How did this happen?
1. With what frequency should I expect it to happen again?
1. How can I recover from it?

Fortunately, there wasn't much in my little test wiki to begin with, so it's not a problem if I lose it entirely, but—unless it occurs very *in*frequently, and recovery is trivial— I'd hate to see this happen to a wiki containing stuff I actually care about. `:-}

**Comment by Jermolene** [comment]
Hi @evadeflow looking at the listing of your `tiddlers` directory, I'm suspicious of the file `$__core.json.tid`; can try removing it?

**Comment by evadeflow** [comment]
Yes—removing that file made the *Internal Javascript Error* go away!  In case you wanted to know what was in it, I've posted a copy [here](https://gist.github.com/evadeflow/7d4e17d2c76e0151775567756e8cc6b2). Thanks!

**Comment by Arlen22** [comment]
@Jermolene is this a result of using `--load`? It almost seems like load shouldn't do that...but maybe it should. WAIT!!! That means that you can load a TW5 file into a directory and all you would need is the boot directory, right? 😁

**Comment by Jermolene** [comment]
Hi @evadeflow 

>  I've posted a copy here

OK, it's a copy of the 5.1.13 core plugin. Under Node.js you're probably running 5.1.14-prerelease. That means that the 5.1.14-prerelease core is being overwritten by the 5.1.13 plugin, causing incompatibilities with other plugins -- the original error message was the tiddlyweb plugin complaining about not seeing a core component that it expected.

Hi @Arlen22 

> is this a result of using --load?

The --load command rejects plugins that are older (by version number) than an already installed plugin. In my tests, the core plugin was rejected when running a 5.1.14-prerelease wiki under Node.js and loading a 5.1.13 standalone wiki via `--load`.

**Comment by Jermolene** [comment]
> WAIT!!! That means that you can load a TW5 file into a directory and all you would need is the boot directory, right? 

Hi @Arlen22 yes it's an intriguing possibility. I've done a little research to try to find a way of creating an `.html` file that is also capable of being run under Node.js (a so-called chameleon files). Without it, we'd need some sort of TiddlyWiki specific .JS stub to load the HTML file and extract the JS.

**Comment by tobibeer** [comment]
in general, though, does it make much sense or rather shouldn't it be discouraged to use the `--load` and `--server` commands in one piped expression?

**Comment by Jermolene** [comment]
Hi @tobibeer 

> in general, though, does it make much sense or rather shouldn't it be discouraged to use the --load and --server commands in one piped expression?

I think @evadeflow ran the commands separately. But anyhow, they are not mutually exclusive in any sense.

**Comment by evadeflow** [comment]
Actually, I did run them singly, as:

```
$ ./bin/node-linux64 TiddlyWiki5/tiddlywiki.js --load ./empty.html --server
```
I'm not sure whether what I'm trying to do makes sense in the context of the TW5 ecosystem, but I'm trying to create a "clone & go" solution for managing my wiki(s) from multiple machines. If you're interested, you could try running these commands on a Windows machine to get a sense of what I'm after:

```
$ git clone --recursive https://github.com/evadeflow/tw5-bootstrap.git
$ cd tw5-bootstrap
$ ./bin/node.exe ./TiddlyWiki5/tiddlywiki.js --server
```

Those commands were run in a 'git bash' shell, so the commands needed on Linux look much the same—just change the last command to:

```
$ ./bin/node-linux64 ./TiddlyWiki5/tiddlywiki.js --server
```

After running that, I'm able to hit `localhost:8080` in a browser, make changes, and sync them using git. And *only* git. (Since all of my machine configurations are managed using git, this is a highly desirable property for me...)

**Comment by Jermolene** [comment]
Hi @evadeflow thanks for the background.

There's no intrinsic problem with your combination of `--load` and `--server` commands. But I still don't understand why the core plugin in the HTML file wasn't rejected.

**Comment by Arlen22** [comment]
I think what happened was he used npm install when he generated the directory, then he downloaded the master and used the same data directory.

**Comment by Arlen22** [comment]
@evadeflow does that sound right?

@jermolene from my tests the tiddlyweb and filesystem plugins do not get loaded.

**Comment by evadeflow** [comment]
> @evadeflow does that sound right?
                                                                                        
Yeah, I think that's essentially what happened: I ran `tiddlywiki --load` with an `npm install`-ed version of TW5, on a monolithic HTML file I'd originally downloaded using the *Download Empty* button on tiddlywiki.com. Once 'deflated', I checked all the individual tiddlers into a local git repo and brought in TW5@master using `git submodule add`. The details are a little fuzzy, but I'm sure at least two—and possibly even *three*—versions of TiddlyWiki were involved while I was tinkering with this stuff.
